### DISCUSSION TITLE
How to do background "cron" jobs at specific times?

### DISCUSSION THREAD

**Discussion by jskherman** [zauberzeug/nicegui#3197] [Q&A] [answered]
### Question

Hi, I'm looking to switch from [Streamlit](https://streamlit.io) and [Holoviz Panel](https://panel.holoviz.org/index.html) but I was wondering how to schedule "cron" jobs or tasks that are executed at a specific time at regular intervals (usually relative to UTC). Is there a way to do this and not via the naïve way of using `ui.timer`? I was looking to refresh data for plots/charts at specific times of the day (e.g. once/twice/etc. per day).

There's [`pn.state.schedule_task()` in Holoviz Panel](https://panel.holoviz.org/how_to/callbacks/schedule.html) that can schedule global tasks and, with the help of the `croniter` library, the time a task is executed can be specified using a cron expression. And from the last I checked, Streamlit has no equivalent for this yet? Is there a similar way to do this in NiceGUI?

**Comment by falkoschindler** [comment, marked as answer]
Hi @jskherman,

I quickly searched for cron-like Python libraries and found [schedule](https://github.com/dbader/schedule). This allows to schedule repeating actions like this:

```py
import schedule
from nicegui import ui

schedule.every(3).seconds.do(lambda: ui.notify('Hello, world!'))

ui.timer(0.1, schedule.run_pending)

ui.run()
```

Here I'm using `ui.timer` to regularly run pending tasks. You could use an infinite loop that is started via `app.on_startup`:
```py
async def loop():
    while True:
        schedule.run_pending()
        await asyncio.sleep(0.1)

app.on_startup(loop)
```

**Reply by jskherman** [reply]
I see! Thank you so much for the reply. Would doing this loop at the interval of `0.1`, `1`, or similar intervals not impact responsiveness or performance by a lot (or at least not annoying enough to bother)?

And am I getting this right that you could do the following with coroutines with respect to `nicegui` and the `schedule` libraries (at least based on [this attempt for documentation](https://github.com/dbader/schedule/pull/391/files))? 

```python
import schedule
from nicegui import ui

async def refresh_data():
    print("Call APIs or database for new query...")
    await asyncio.sleep(3) # substitute for expensive computation/data cleaning

schedule.every().day.at("20:00", "Europe/London").do(lambda: asyncio.create_task(refresh_data()))

async def loop():
    while True:
        schedule.run_pending()
        await asyncio.sleep(0.1)

app.on_startup(loop)
```

**Reply by falkoschindler** [reply]
I assume `schedule.run_pending()` to be pretty cheap if there is nothing to do. So calling it once or ten times per second shouldn't be a problem.

Yes, calling `asyncio.create_task` should work well.

**Reply by jskherman** [reply]
Thank you for your answer. Looking forward to using NiceGUI!
